File: Deck Shuffling/C Sharp Shuffling/Cards/Deck.cs
/// File: Deck.cs
/// Programmer: Philip Kempton
/// First Version: March 23rd, 2020
/// Description: This file contains the Deck class

using System.Collections.Generic;
using System.Text;

namespace Cards
{
    /// <summary>
    /// This class represents a deck of playing cards
    /// (Ace through King of each standard suit)
    /// </summary>
    public class Deck
    {
        /// <summary>
        /// Name of the deck of cards
        /// </summary>
        private string Name { get; set; }

        /// <summary>
        /// The list of cards in the deck
        /// </summary>
        private List<Card> Cards { get; set; }

        /// <summary>
        /// Array of all valid suits (used for Population() method)
        /// </summary>
        private static readonly string[] SUITS = {
                "Spades", "Hearts", "Diamonds", "Clubs"
        };

        /// <summary>
        /// Array of all valid card names (used for Population() method)
        /// </summary>
        private static readonly string[] CARD_NAMES = {
                "Ace", "Two", "Three", "Four", "Five", "Six", "Seven",
                "Eight", "Nine", "Ten", "Jack", "Queen", "King"
        };

        /// <summary>
        /// Constructor
        /// Populates the list of cards and optionally shuffles them
        /// </summary>
        /// <param name="name">bool to decide if deck should start shuffled</param>
        public Deck(string name)
        {
            this.Name = name;
            this.Populate();
        }

        /// <summary>
        /// Populates the card list with Ace through King of all four suits
        /// </summary>
        private void Populate()
        {
            // reset the deck list
            this.Cards = new List<Card>();

            // create all cards from ace -> king of each suit and put into cards list 
            foreach (string suit in SUITS)
            {
                foreach (string cardName in CARD_NAMES)
                {
                    this.Cards.Add(new Card(suit, cardName));
                }
            }
        }

        /// <summary>
        /// Shuffles the cards in the cards list
        /// </summary>
        public void Shuffle(int numOfShuffles)
        {
            // this part is fun, 
            // created a List Extension method that implements a shuffling algorithm
            this.Cards.Shuffle(numOfShuffles);
        }

        /// <summary>
        /// OVERRIDE
        /// Creates and returns string representation of the deck
        /// </summary>
        /// <returns>list of cards in deck</returns>
        public override string ToString()
        {
            // put name of deck on it's own line, follow by an empty line
            // then list each card in the deck on it's own line
            StringBuilder sb = new StringBuilder();
            sb.AppendLine(string.Format("{0}\n", this.Name));
            int lastIndex = this.Cards.Count - 1;
            foreach (Card card in this.Cards)
            {
                if (card != this.Cards[lastIndex])
                {
                    sb.AppendLine(card.ToString());
                }
            }
            sb.Append(this.Cards[lastIndex]);
            return sb.ToString();
        }
    }
}

File: C_Shuffling/deck.c
// File: deck.c
// Programmer: Philip Kempton
// First Version: March 24th, 2020
// Decription: Contains Main() for this application.

#include "deck.h"

// Name: deckToString
// Description: This function builds and returns a string representation of a deck
// Parameters: 
//		d:	the deck
// Returns: string of all cards in the deck
char* deckToString(deck d)
{
	// Define card suits
	char suits[SUIT_COUNT][MAX_SUIT_LEN] = {
		"Spades", "Hearts", "Diamonds", "Clubs"
	};

	// Define card names
	char cardNames[VALUE_COUNT][MAX_VALUE_LEN] = {
		"Ace", "Two", "Three", "Four", "Five", "Six", "Seven", "Eight", "Nine", "Ten", "Jack", "Queen", "King"
	};

	// Define string to store string representation of the deck
	char deckStr[DECK_SIZE * CARD_STR_LEN] = { 0 };

	// Define string to store string representation of a card
	char cardStr[CARD_STR_LEN] = { 0 };

	// Loop through each "card" in the deck, turning it into its' string representation and concatenating it to the deck string
	// This is done with modulo and division to turn each char value (0-51) in a specific card
	for (int i = 0; i < DECK_SIZE; i++)
	{
		int valueIndex = d.cards[i] % VALUE_COUNT;
		int suitIndex = d.cards[i] / VALUE_COUNT;
		sprintf_s(cardStr, CARD_STR_LEN, "%s of %s\n", cardNames[valueIndex], suits[suitIndex]);
		strncat_s(deckStr, DECK_SIZE * CARD_STR_LEN, cardStr, strlen(cardStr));
	}
	return deckStr;
}

// Name: newDeck
// Description: Builds and returns a deck, ordered deck
// Parameters: n/a
// Returns: new deck
deck newDeck()
{
	// assign each "card" with a unique value, from 0-51 (to represent the different cards)
	deck d;
	for (int i = 0; i < DECK_SIZE; i++)
	{
		d.cards[i] = i;
	}
	return d;
}

// Name: newDeck
// Description: Shuffles a deck using a randomization function
// Parameters: 
//		d:			the deck to shuffle
//		randFunc:	function pointer to a rand function
// Returns: n/a
void shuffleDeck(deck* d, int (*randFunc)())
{
	shuffle(d->cards, DECK_SIZE, sizeof(char), randFunc);
}